File: growth/oj/leet_code/algorithms/322-coin-change.py
# -*- coding: utf-8 -*-

'''
    TIPS: 递推方程很简单, dp[i]代表构成i元需要的硬币数量, 需要指出的是当无法构成i元的时, dp[i] = 0
    TODO:
        1. 内存上的优化
        2. 速度上的优化
'''

class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if len(coins) == 0:
            return -1
        if amount == 0:
            return 0

        dp = [0] * (amount+1)

        i = 1
        while i <= amount:
            for x in coins:
                if i - x >= 0:
                    if (i - x == 0) or (dp[i-x]>0):
                        dp[i] = min(dp[i-x]+1, dp[i]) if dp[i] > 0 else dp[i-x]+1
            i += 1
        #print dp
        return dp[amount] if dp[amount] > 0 else -1

if __name__ == '__main__':
    obj = Solution()
    print obj.coinChange([1,2,5], 11)
    print obj.coinChange([2], 3)
    print obj.coinChange([1], 0)
    pass

File: growth/oj/leet_code/algorithms/322-coin-change.cpp
#include<iostream>
#include<vector>

using namespace std;
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount <= 0) return 0;
        if (coins.size() == 0) return -1;
        vector<int> dp = vector<int>(amount+1);
        for (int i = 1; i <= amount; i++) {
            dp[i] = -1;
            for (int j = 0; j < coins.size(); j++) {
                if (i - coins[j] >= 0) {
                    int ans = i - coins[j] == 0 ? 1 : dp[i];
                    if (dp[i-coins[j]] > 0)
                        ans = dp[i] == -1 ? dp[i-coins[j]] + 1 : min(ans, dp[i-coins[j]] + 1);
                    dp[i] = ans;
                }
            }
        }
        return dp[amount];
    }
};

int main() {
    Solution s;
    vector<int> v{1,2,5};
    cout << s.coinChange(v, 11) << endl;
    v = {2};
    cout << s.coinChange(v, 3) << endl;
    return 0;
}
